blocktx/client.go

```diff
@@ -26,55 +27,100 @@ type ClientI interface {
 	GetLastProcessedBlock(ctx context.Context) (*blocktx_api.Block, error)
 	GetMinedTransactionsForBlock(ctx context.Context, blockAndSource *blocktx_api.BlockAndSource) (*blocktx_api.MinedTransactions, error)
+	Shutdown()
 }
 
+const (
+	logLevelDefault      = slog.LevelInfo
+	retryIntervalDefault = 10 * time.Second
+)
+
 type Client struct {
-	address string
-	logger  utils.Logger
-	conn    *grpc.ClientConn
-	client  blocktx_api.BlockTxAPIClient
+	logger                *slog.Logger
+	client                blocktx_api.BlockTxAPIClient
+	retryInterval         time.Duration
+	retryTicker           *time.Ticker
+	shutdownCompleteStart chan struct{}
+	shutdown              chan struct{}
 }
 
-func NewClient(l utils.Logger, address string) ClientI {
-	btc := &Client{
-		address: address,
-		logger:  l,
+func WithLogger(logger *slog.Logger) func(*Client) {
+	return func(p *Client) {
+		p.logger = logger.With(slog.String("service", "btx"))
 	}
+}
 
-	conn, err := btc.dialGRPC()
-	if err != nil {
-		btc.logger.Fatalf("Failed to connect to block-tx server at %s: %v", btc.address, err)
+func WithRetryInterval(d time.Duration) func(*Client) {
+	return func(p *Client) {
+		p.retryInterval = d
 	}
-	btc.logger.Infof("Connected to block-tx server at %s", btc.address)
+}
 
-	btc.conn = conn
-	btc.client = blocktx_api.NewBlockTxAPIClient(conn)
+type Option func(f *Client)
+
+func NewClient(client blocktx_api.BlockTxAPIClient, opts ...Option) ClientI {
+	btc := &Client{
+		logger:                slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: logLevelDefault})).With(slog.String("service", "btx")),
+		client:                client,
+		retryInterval:         retryIntervalDefault,
+		shutdown:              make(chan struct{}, 1),
+		shutdownCompleteStart: make(chan struct{}, 1),
+	}
+
+	for _, opt := range opts {
+		opt(btc)
+	}
 
 	return btc
 }
 
 func (btc *Client) Start(minedBlockChan chan *blocktx_api.Block) {
-	for {
-		stream, err := btc.client.GetBlockNotificationStream(context.Background(), &blocktx_api.Height{})
-		if err != nil {
-			btc.logger.Errorf("Error getting block notification stream: %v", err)
-		} else {
-			btc.logger.Infof("Connected to block-tx server at %s", btc.address)
-
-			var block *blocktx_api.Block
-			for {
-				block, err = stream.Recv()
+	btc.retryTicker = time.NewTicker(btc.retryInterval)
+
+	go func() {
+		defer func() {
+			btc.shutdownCompleteStart <- struct{}{}
+		}()
+
+		for {
+			select {
+			case <-btc.retryTicker.C:
+				stream, err := btc.client.GetBlockNotificationStream(context.Background(), &blocktx_api.Height{})
 				if err != nil {
-					break
+					btc.logger.Error("failed to get block notification stream", slog.String("err", err.Error()))
+					continue
 				}
 
-				btc.logger.Infof("Block %s\n", utils.ReverseAndHexEncodeSlice(block.Hash))
-				utils.SafeSend(minedBlockChan, block)
+				btc.logger.Info("Connected to block-tx server")
+
+				var block *blocktx_api.Block
+				for {
```

Comment:
I'm not sure nested channel processing inside another select is good idea.

Reply:
Can you explain your concern in more detail?
The first select waits for either the ticker or the shutdown to be triggered
The second select waits for either the receiving of the block to end or also the shutdown to be triggered.

In this way the `Shutdown()` function waits until `block, err = stream.Recv()` is finished.

Reply:
1. I've refactored the client start test in this branch https://github.com/bitcoin-sv/arc/blob/b182300b6efe04c9d7a358fb39b5cdaaf0876a99/blocktx/client.go#L70 

with single select. The result is exact same and much less confusing than nested selects.